IPIP/0373-double-hash-dht.md

```diff
@@ -0,0 +1,404 @@
+# IPIP 0000: Double Hash DHT
+
+![wip](https://img.shields.io/badge/status-wip-orange.svg?style=flat-square)
+- DRI: [Guillaume Michel](https://github.com/guillaumemichel)
+- Start Date: 2023-01-18
+- Related Resources:
+  - [Specs in Notion](https://pl-strflt.notion.site/Double-Hashing-for-Privacy-ff44e3156ce040579289996fec9af609)
+  - [WIP Implementation](https://github.com/ChainSafe/go-libp2p-kad-dht)
+  - https://github.com/ipfs/specs/pull/334
+  - https://github.com/ipfs/specs/issues/345
+
+## Summary
+
+This IPIP contains the up-to-date Spec of the IPFS Double Hash DHT. The Double Hashing DHT aims at providing some Reader Privacy guarantees to the IPFS DHT.
+
+This document is still WIP, all feedback is more than welcome. Make sure to write your thoughts about the [open questions](#open-questions) in the PR.
+
+## Table of Contents
+
+1. [Motivation](#motivation)
+2. [Detailed Design](#detailed-design)
+3. [Design Rationale](#design-rationale)
+4. [User benefits](#user-benefits)
+5. [Migration](#migration)
+6. [Threat Model](#threat-model)
+7. [Alternatives](#alternatives-for-dht-reader-privacy)
+8. [Open Questions](#open-questions)
+
+## Motivation
+
+IPFS is currently lacking of many privacy protections. One of its principal weaknesses currently lies in the lack of privacy protections for the DHT content routing subsystem. Currently in the IPFS DHT, neither readers (clients retrieving content) nor writers (hosts storing and distributing content) have much privacy with regard to content they consume or publish. It is trivial for a DHT server node to associate the requestor's identity with the accessed content during the routing process. A curious DHT server node, can request the same CIDs to find out what content other users are consuming. Improving privacy in the IPFS DHT has been a strong request from the community for some time.
+
+The changes described in this document introduce a DHT privacy upgrade boosting the reader’s privacy. It will prevent DHT tracking as described above, and add Provider Records Authentication. The proposed modifications also add a slight Writer Privacy improvement as a side effect.
+
+## Detailed Design
+
+**Magic Values**
+- bytes("CR_DOUBLEHASH")
+- bytes("CR_SERVERKEY")
```

Comment:
You should pad (zero or nothing to hide numbers) magic values, this allows implementations to cache the magic vector instead of having to handle with block padding and remainder.
```suggestion
- bytes("CR_DOUBLEHASH\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00")
- bytes("CR_SERVERKEY\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00")
```